src/ai/backend/manager/models/minilang/queryfilter.py

```diff
@@ -0,0 +1,186 @@
+            ''')
+        tree = self._JSON_PARSER.parse(query)
+        return tree
+
+    def _check_type(self, tree, type=""):
+        if type == 'word':
+            if type(tree) == str:
+                return True
+        elif tree.data == type:
+            return True
+        return False
+
+    def _token2str(self, token):
+        if type(token) == list:
+            token = token[0]
+        assert type(token) == lark.lexer.Token
+        return str(token)
+
+    def tree_partition(self, tree, ret_table=True, ret_filter=True, ret_select=True):
+        ret = []
+        if tree.data == 'object':
+            tree = tree.children
+        if ret_table:
+            table = tree.children[0].children
+            table = self._token2str(table)
+            ret.append(table)
+        if ret_filter:
+            filter = tree.children[1]
+            ret.append(filter)
+        if ret_select:
+            select = tree.children[2]
+            ret.append(select)
+        if len(ret) == 1:
+            return ret[0]
+        return ret
+
+    def tree2table(self, tree):
+        table = self.tree_partition(
+            tree, ret_table=True, ret_filter=False, ret_select=False)
+        return table
+
+    def tree2where(self, tree):
+        filter_stack = []
+        count_stack = []
+        bracket_stack = []
+        where = ""
+        if tree.data != "filter":
+            tree = self.tree_partition(
+                tree, ret_table=False, ret_filter=True, ret_select=False)
+        tree = list(tree.iter_subtrees_topdown())
+        for i in range(len(tree)):
+            children = tree[i].children
+            if type(children[0]) == lark.lexer.Token:
+                token = self._token2str(children)
+                if token in self._FILTERS:
+                    filter_stack.append(token)
+                    if token == 'and' or token == 'or':
+                        where += "("
+                        count_stack.append(0)
+                    elif count_stack:
+                        count_stack[-1] += 1
+                        while count_stack:
+                            if count_stack[-1] == 2:
+                                bracket_stack.append(") ")
+                                count_stack.pop()
+                                if count_stack:
+                                    count_stack[-1] += 1
+                            else:
+                                break
+                else:
+                    where += token + " "
+                    if tree[i - 1].data != "pair":
+                        while bracket_stack:
+                            where = where[:-1]
+                            where += bracket_stack.pop()
+                    if filter_stack:
+                        where += self._FILTERS[filter_stack.pop()] + " "
+        if where[0] == "(" and where[-2] == ")":
+            where = where[1:-1]
+        return where[:-1]
```

Comment:
I think it would be better to generate the column object representation directly instead of generating the `where` string. Let's think about how we could start with a root expression, maybe `sa.sql.expression.true`.